PLAT-1340:  Add selectable highlighted

mihaelpalfi opened the pull request:
[JIRA ticket](https://reciprocitylabs.atlassian.net/browse/PLAT-1340)

#### Description
As per JIRA ticket

- Additionally added that child rows get selected and checked also.
- Additionally added that on table checkbox check all rows get selected.
- indeterminate state will be added in next pr
- Added additional story for tree with header and updated stories.
- Additional unit tests will be done in next pr.

![image](https://user-images.githubusercontent.com/59830865/110375305-16048280-8052-11eb-9ba5-e59bd2435449.png)


#### ChangeLOG

[Changelog](https://zen-ui.zengrc.com/?path=/docs/changelog--page)


#### Useful links
[Zen UI readme](https://github.com/reciprocity/zen-ui/blob/main/README.md)

[PR review guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/pull_request_reviews.md)

[Git commit guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/commit_guidelines.md)

chavlji commented:
A few bugs that should be resolved before diving into code review:

1. Child rows shouldn't be expanded when the parent row is expanded.
2. Collapsed child node's background is white, even though it's not a root row.
https://user-images.githubusercontent.com/5729421/111273553-417c0400-8634-11eb-93d5-19c27e06411f.mov

3. Parent's checkbox doesn't get updated when children selected state changes. I don't think this is a task for this PR, so please create a ticket and add a link below.
https://user-images.githubusercontent.com/5729421/111273925-ac2d3f80-8634-11eb-8c35-a8c9638d33b0.mov

4. Is this spacing really ok? Seems like gap between dropdown arrow and label is really big (1.)
5. Spacing between checkbox and label is even bigger in case row has no children (2.). Please talk to designers how this should be done.
![Screenshot 2021-03-16 at 08 51 31](https://user-images.githubusercontent.com/5729421/111274432-468d8300-8635-11eb-9953-a8507f19d27f.jpg)

chavlji reviewed (changes requested):
A few bugs that should be resolved before diving into code review:

1. Child rows shouldn't be expanded when the parent row is expanded (video under pt. 2 attached).
2. Collapsed child node's background is white, even though it's not a root row (video attached).
https://user-images.githubusercontent.com/5729421/111273553-417c0400-8634-11eb-93d5-19c27e06411f.mov

3. Parent's checkbox doesn't get updated when children selected state changes. I don't think this is a task for this PR, so please create a ticket and add a link below.
https://user-images.githubusercontent.com/5729421/111273925-ac2d3f80-8634-11eb-8c35-a8c9638d33b0.mov

4. Is this spacing really ok? Seems like gap between dropdown arrow and label is really big (Pt. 1 on image under pt. 5)
5. Spacing between checkbox and label is even bigger in case row has no children (Pt. 2 on image under pt. 5). Please talk to designers how this should be done.
![Screenshot 2021-03-16 at 08 51 31](https://user-images.githubusercontent.com/5729421/111274432-468d8300-8635-11eb-9953-a8507f19d27f.jpg)

mihaelpalfi commented:
> A few bugs that should be resolved before diving into code review:

1. Fixed the children expanding issue.
2. Fixed together
3. Next PR for intermediate state.
4. Fixed together
5. Fixed together

chavlji commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+    return children;
+  }
+
+  descendants(target: HTMLZenTableRowElement): HTMLZenTableRowElement[] {
```
rename to `rows`

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -8,38 +8,38 @@ import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
+  /** Row expanded */
+  @Event() rowExpanded: EventEmitter<boolean>;
 
   children(): HTMLZenTableRowElement[] {
```
Shouldn't this be named `descenants`? Children are direct children, so outDepth+1. Descendants are items inside this row on any depth.

mihaelpalfi replied:
Hmm this are direct children??

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -8,38 +8,38 @@ import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
-    let next = this.element.nextElementSibling as HTMLZenTableRowElement;
+    let next = this.host.nextElementSibling as HTMLZenTableRowElement;
 
     // Get all rows that have depth greater then the parent
```
Invalid comment if this function returns children.

mihaelpalfi replied:
removed

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -80,60 +64,56 @@ export class ZenTableRow {
+    this.rowExpanded.emit(!this.expanded);
+  }
+
+  onSelect(): void {
```
Is this triggered if you call `document.querySelector('#my-row').selected = true` somewhere in my code?

mihaelpalfi replied:
Added new PLAT-1541

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -80,60 +64,56 @@ export class ZenTableRow {
     return !!this.children().length;
   }
 
+  onExpand(): void {
```
This isn't triggered if you call `document.querySelector('#my-row').expanded = true` somewhere in the code.

mihaelpalfi replied:
Added new PLAT-1541

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -8,38 +8,38 @@ import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
+  @Prop() readonly expanded: boolean = false;
 
   /** Is cell full span (colspan=number of cells) */
   @Prop() readonly fullSpan = false;
```
Where is this used?

mihaelpalfi replied:
You are right only in the css. Removed

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -8,38 +8,38 @@ import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
   @Prop() readonly fullSpan = false;
 
   /** Depth position of row (read-only) */
   @Prop() readonly depth: number = 0;
```
Check that it works correctly if you call `document.querySelector('#my-row').depth = 15` somewhere from eg. some button click handler. I think you need watchers for most of these props.

mihaelpalfi replied:
I added property  mutable to expanded and selected. I needed it because of the change addressed up.

chavlji commented on src/components/zen-table-row/test/zen-table-row.spec.tsx:
```diff
@@ -10,4 +10,16 @@ describe('zen-table-row', () => {
     expect(page.root.innerHTML).toEqual('Content');
   });
+
+  it('should render checkbox', async () => {
```
Add more tests. Current coverage is 69%.

mihaelpalfi replied:
This is a subtask PLAT-1537

chavlji replied:
Haven't we all agreed that tests should be written in the same pr? If we don't agree, we have to discuss it again. In this particular case, do it in a separate pr.

mihaelpalfi replied:
Didnt know we agreed to this. When?

chavlji commented on src/components/zen-table-row/zen-table-row.stories.mdx:
```diff
@@ -0,0 +1,151 @@
+
+<br />
+
+### 1. Single row
```
You can remove those stories. Keep just story with controls. All of these stories are alreday presented in the table stories page.

mihaelpalfi replied:
Dont agree. The stories should contain the component possibilities. In the table component they are all joined together.

chavlji replied:
Ok. From story 2, please remove header and keep just one row. What I've meant is that focus here should be on one row only so it's not confused with the table. How table header and selectable row look like is already presented in the table story page.

mihaelpalfi replied:
Done.

chavlji commented on src/components/zen-table-row/zen-table-row.stories.mdx:
```diff
@@ -0,0 +1,151 @@
+argTypes['expanded'].table.disable = true;
+argTypes['visible'].table.disable = true;
+
+<Meta title="Tables/Table Row" component="zen-table-row" argTypes={argTypes} />
```
Add all table components stories to `Lists/Table/Row`, `Lists/Table/Header`, `Lists/Table/Table`

chavlji commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -80,60 +64,56 @@ export class ZenTableRow {
-    };
-    const widgetClass = {
-      widgets: true,
+      expandable: this.expandable,
```
As mentioned in previous Prs... use `reflect` and attribute selectors and don't tranform props into classes.

mihaelpalfi replied:
Uff if i start changing this on all files i have work for 2 days :) Because css will stop working and like you know i will struggle. I can create a new ticket maybe.

chavlji replied:
Add new ticket then and paste link here.

mihaelpalfi replied:
PLAT-1548

chavlji reviewed (changes requested):
Getting there. It works much better now!

But this PR is waaay too long! 750 lines. It's impossible to review carefully. Description of what this code does is scattered all over the place (a little in jira, a bit in pr description, a few bulletins in comments...). Next time please chop it into smaller PRs.

One thing to fix:
1. Popovers are covered with the next rows
![Screenshot 2021-03-17 at 12 54 35](https://user-images.githubusercontent.com/5729421/111463703-28508180-8720-11eb-90ab-43263cfead85.jpg)
This happens because you have `transform: scale(1)` on table row. But if you don't have it, row doesn't have `position: relative` and you can't position elements absolutely... It's quite a big problem.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+    return children;
+  }
+
+  descendants(target: HTMLZenTableRowElement): HTMLZenTableRowElement[] {
```
Done

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+    return descendants;
+  }
+
+  @Listen('rowSelected')
```
Done

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+    if (expanded) {
+      this.children(target).forEach(n => {
+        n.visible = true;
+        n.classList.add('expanded-bg');
```
It was there but then it didn work correctly. The expanded inner child lost the background on close.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+  @Element() host: HTMLZenTableElement;
+
+  expandedChanged(target: HTMLZenTableRowElement, expanded: boolean): void {
+    target.expanded = expanded;
```
Moved the logic for setting expandable and selectable to row and header. Also cleaned up the handler part.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+  }
+
+  selectedChanged(target: HTMLZenTableRowElement, selected: boolean): void {
+    target.selected = selected;
```
Changed.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+  }
+
+  headerSelectedChanged(target: HTMLZenTableHeaderElement, selected: boolean): void {
+    target.selected = selected;
```
changed.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
+  headerSelectedChanged(target: HTMLZenTableHeaderElement, selected: boolean): void {
+    target.selected = selected;
+
+    let next = target.nextElementSibling as HTMLZenTableRowElement;
```
Done

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -7,26 +7,44 @@ import { applyPrefix } from '../helpers/helpers';
+  /** Row selected */
+  @Event() headerSelected: EventEmitter<boolean>;
+
+  rows(): HTMLZenTableRowElement[] {
```
Hmm you suggested to be rows. Going in circle. :) In my opinion and how Leo mentioned on your pr it should be getRows.

chavlji replied:
ok

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -7,26 +7,44 @@ import { applyPrefix } from '../helpers/helpers';
+    return this.rows().some(row => row.selected);
+  }
+
+  allSelectedRows(): boolean {
```
Thats why you use "get" in the names returning something and here "has". But again this variable name  was your suggestion. :) Changed to hasAllRowsSelected, hasRowsSelected, hasExpandableRows.

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -38,26 +56,43 @@ export class ZenTableHeader {
   }
 
+  onSelect(): void {
+    this.headerSelected.emit(!this.selected);
```
Yes changed. Before i had issues doing it like a toggle. But i think that was another issue that resolved it now.

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -38,26 +56,43 @@ export class ZenTableHeader {
     forEach(elements, setSticky);
```
Done

mihaelpalfi commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -80,60 +64,56 @@ export class ZenTableRow {
   }
 
+  onExpand(): void {
+    this.rowExpanded.emit(!this.expanded);
```
Yes done

mihaelpalfi commented on src/components/zen-table-row/zen-table-row.tsx:
```diff
@@ -8,38 +8,38 @@ import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
-    }
-  }
+  /** Row selected */
+  @Event() rowSelected: EventEmitter<boolean>;
```
Changed to rowSelectChange, rowExpandChange

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -7,26 +7,44 @@ import { applyPrefix } from '../helpers/helpers';
+    return this.rows().every(row => row.selected);
   }
 
   setSticky(): void {
```
Is this not handled if the prop is mutable?

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -38,26 +56,43 @@ export class ZenTableHeader {
+    this.observer.observe(table, {
+      childList: true,
+      attributes: true,
+      subtree: true,
```
Removed. Still works.

mihaelpalfi commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,94 @@
 export class ZenTable {
+  @Element() host: HTMLZenTableElement;
+
+  expandedChanged(target: HTMLZenTableRowElement, expanded: boolean): void {
```
This is a separate pr.

mihaelpalfi commented on src/components/zen-table-header/zen-table-header.tsx:
```diff
@@ -7,26 +7,44 @@ import { applyPrefix } from '../helpers/helpers';
+  @Prop() readonly selectable: boolean = false;
+
+  /** Select all rows */
+  @Prop() readonly selected: boolean = false;
```
Added new PLAT-1541

chavlji commented on src/components/zen-table/zen-table.tsx:
```diff
@@ -1,11 +1,93 @@
+    if (expanded) {
+      this.children(target).forEach(n => {
+        n.visible = true;
+        n.classList.add('expanded-bg');
```
Css must have been wrong then. Try something like this:
`:host([depth]:not([depth=0])), :host(.expanded) { background-color: $gray; }`

mihaelpalfi replied:
Done.

chavlji reviewed (changes requested):
A few minor changes left, then we can merge.

**Regarding this:**
_Thats why you use "get" in the names returning something and here "has". But again this variable name was your suggestion. :) Changed to hasAllRowsSelected, hasRowsSelected, hasExpandableRows._

That's not true. Check git blame. It's your code.
The point is that in English you ask questions like: "Are all rows selected?" and not "Are all selected rows?". Second question has a different meaning which means function name is not describing what the function actually does, which is a bad practice.
In your new change word order is ok.
